### ISSUE TITLE
Download all the pages in the same time

### ISSUE THREAD

**Issue by DyeffersonAz** [Xonshiz/comic-dl#123]
Why isn't this possible?

If it is possible: there could be an argument to a person that really wants it.

NOTE: I'm a real dummie on it. XD

**Comment by dsanchezseco** [comment]
It's doable.... although the 90% of the time is spent on fooling cloudfare so the download rate is slightly improved. 

I'd tried with Thanos-2016 from readcomiconline.to  (18 issues ~ 23 pages each) and the time was **374.815049887 with four threads** -> four pages in parallel.

In single thread, **one page at a time the time was 428.929593086** 

The required changes are 

globalFunctions.py line 45
```
def downloader(self, image_and_name, referer, directory_path, **kwargs):
        self.logging = kwargs.get("log_flag")
        image_ddl = image_and_name[0] 
        file_name = image_and_name[1]
```
and readcomicOnlineto.py line 83

```
        pool = ThreadPool(4)
        pool.map(partial(globalFunctions.GlobalFunctions().downloader, referer=comic_url, directory_path=directory_path), zip(links,file_names))

    globalFunctions.GlobalFunctions().conversion(directory_path, conversion, delete_files, chapter_number)
```

as well as the required imports and declarations.

If it's relevant @Xonshiz i can try to further develop this concurrency and do a PR.

**Comment by DyeffersonAz** [comment]
I mean: it could be an argument.

**Comment by Xonshiz** [comment]
@dsanchezseco it would be great if you could look into this when you get time. I've been a bit too busy lately with college and stuff. And yes, that download bar needs to be fixed somehow xD! Maybe remove it altogether? but, it gives a good visual :P

**Comment by dsanchezseco** [comment]
Cool, I'll try to work on it this weekend. :)

**Comment by DyeffersonAz** [comment]
:zap:

**Comment by dsanchezseco** [comment]
![image](https://user-images.githubusercontent.com/20790609/41678005-2c5759fe-74ca-11e8-8524-590585b5c877.png)

You know we do this in our free/spare time, don't you?

**Comment by kidpixo** [comment]
I was thinking about tinkering witht the download bar, instead a progress for single page a progress for each chater could be easyier.

Should I open a separate issue on the progress bar?

Diving in the code I understood each single manga has its parser and then uses common functions from `globalFunctions.pya ,  in particular `downloader` is also spitting out the progress bar.

There a lot of places where information are printed to the screen anyway, this could almost entirally removed and use a more informative progress indicator, I think.

I use a lot [tqdm/tqdm: A fast, extensible progress bar for Python and CLI](https://github.com/tqdm/tqdm) a lot lately, is really great and you can manipulate the progress bar and nest really easily.

In our case you can simply do 

```python
# code on the manga, i.e. generate the chapter list
for i in tqdm.tqdm.(manga_chapters_list):
    # code on each chapter, i.e. extract each single page link
    for i in tqdm.tqdm.(chapter_pages_list):
        # code on each page, i.e. download them
        globalFunctions.GlobalFunctions().downloader
```

For the parallelization part, I'm using [Embarrassingly parallel for loops — joblib 0.11 documentation](https://pythonhosted.org/joblib/parallel.html#using-the-threading-backend) just because I am really a beginner on this stuff, but if you can wrap your function in a single call and you're ok in getting the results aggreted in a list afterward, it is just one line call.

**Comment by dsanchezseco** [comment]
@kidpixo from my part feel free to mod the code. I'm kind on busy nowadays and I have little time to spare on this, so thanks! Any improvement is welcome.

**Comment by kidpixo** [comment]
@dsanchezseco thanks, seems to work. I don't have much time either, this is a side hobby for something I find useful.

Just a question you all know better : do I need to change each single scraper or I can work only on the `globalFunctions.py` ? 

in the first case I would not try at all, to much work!

![screen shot 2018-07-09 at 09 16 22](https://user-images.githubusercontent.com/2255209/42436007-d423dea4-8358-11e8-8b17-b74406b52c59.jpg)

Developing in Docker :-D

**Comment by dsanchezseco** [comment]
Well i suppose that it depends on how you made it. I think that maybe, as it happen in my case with the auto donwloader you'll have to code the function once in the `globalFuntions.py` for example and then just add it to the different scrappers in a copy/paste fashion(with little mods as some var names change).

**Comment by kidpixo** [comment]
Ok, 
here an example : 

![screen shot 2018-07-09 at 09 53 56](https://user-images.githubusercontent.com/2255209/42437738-47ed65ee-835e-11e8-8ffd-bca2df2c7430.jpg)

The lines are:

1.  a progress bar for the whole manga, that advances each time the second level is done.
2. (below) progress for the chapter,  advances for each single page loaded
3. (below)  progress for the single image, advances for each single chunk downloaded.

I had to mock `globalFunctions.GlobalFunctions().info_printer()` to just pass, it s called in each single scraper, but seems to work and now it uses a lot less screen space!

**Comment by dsanchezseco** [comment]
cool, really cool!

**Comment by kidpixo** [comment]
It was easier than I thought  :-D

Stupid as I am I did the changes in master... 

I'll open a separate issue to discuss this and try to put everything in a new branch.

Last questions : 
- is my impression or all this is still heavilly python 2?
- do you use [Gitter](https://gitter.im/)? you can open a channel for stupid questions (like a couple I have). It is a chat linked to github repos.

Cheers

**Comment by dsanchezseco** [comment]
No idea on any of those, i'm not the owner, ask @Xonshiz

**Comment by kidpixo** [comment]
Right, sorry!

@Xonshiz I created a [develop](https://github.com/kidpixo/comic-dl/tree/develop) in my fork , if you create one here too I can PR a merge.

You can find all the mods there.

Better not start to mess with master with experiments...

**Comment by kidpixo** [comment]
@dsanchezseco  I't trying to move this discussion in #153 , can you add your comments if you have some minutes?

**Comment by dsanchezseco** [comment]
@Xonshiz perfect, reject the current PR and i'll create a new one into develop

**Comment by Xonshiz** [comment]
Yep, sure.

**Comment by Xonshiz** [comment]
Merged and binaries released for the same. Hence, Closing.
